src/ucp/rma/basic_rma.c

```diff
@@ -377,6 +426,32 @@ ucs_status_t ucp_put_nbi(ucp_ep_h ep, const void *buffer, size_t length,
     return status;
 }
 
+ucs_status_ptr_t ucp_put_nb(ucp_ep_h ep, const void *buffer, size_t length,
+                            uint64_t remote_addr, ucp_rkey_h rkey,
+                            ucp_send_callback_t cb)
+{
+    ucp_ep_rma_config_t *rma_config;
+    ucs_status_ptr_t ptr_status;
+    ucs_status_t status;
+
+    UCP_RMA_CHECK_PARAMS_PTR(buffer, length);
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    status = UCP_RKEY_RESOLVE(rkey, ep, rma);
+    if (status != UCS_OK) {
+        ptr_status = UCS_STATUS_PTR(status);
+        goto out_unlock;
+    }
+
+    rma_config = &ucp_ep_config(ep)->rma[rkey->cache.rma_lane];
+    ptr_status = ucp_rma_nonblocking_cb(ep, buffer, length, remote_addr, rkey,
```

Comment:
for short puts, would be more optimal to try sending them immediately before allocating a request (e.g https://github.com/openucx/ucx/blob/master/src/ucp/tag/tag_send.c#L177)

Reply:
Are short puts guaranteed to complete on the target before returning?

Reply:
no, it only guarantees local completion (send buffer is reusable). However, this is the same level of guarantee that request-based completion provides (when request is completed)

Reply:
ah, makes sense. i would at some point like to add a worker option for remote completion.

Reply:
do we need to know about remote completion per-operation (today we have only global one, with flush)? this already came up few times in the past..

Reply:
Its useful to users of the library. I would think it would be easy to implement as a per-worker feature,

Reply:
Or even as a different API call.

Reply:
Since it only affects put.